web_services/foxx/brightid5/schemas.js

```diff
@@ -0,0 +1,257 @@
+const joi = require('joi');
+
+// lowest-level schemas
+var schemas = {
+  score: joi.number().min(0).max(100).default(0),
+  timestamp: joi.number().integer(),
+};
+
+// extend lower-level schemas with higher-level schemas
+schemas = Object.assign({
+  user: joi.object({
+    id: joi.string().required().description('the user id'),
+    score: schemas.score,
+    verifications: joi.array().items(joi.string()),
+    flaggers: joi.object().description("an object containing ids of flaggers as key and reason as value"),
+  }),
+  groupBase: joi.object({
+    id: joi.string().required().description('unique identifier of the group'),
+    members: joi.array().items(joi.string()).required().description('brightids of group members'),
+    type: joi.string().required().description('type of group which is "primary" or "general"'),
+    founders: joi.array().items(joi.string()).required().description('brightids of group founders'),
+    admins: joi.array().items(joi.string()).required().description('brightids of group admins'),
+    isNew: joi.boolean().required().description('true if some of founders did not join the group yet and group is still in founding stage'),
+    score: schemas.score,
+    url: joi.string().required().description('url of encrypted group data (name and photo)'),
+    timestamp: schemas.score.required().description('group creation timestamp'),
+  }),
+  app: joi.object({
+    id: joi.string().required().description('unique app id'),
+    name: joi.string().required().description('app name'),
+    context: joi.string().required().description('app context'),
+    verification: joi.string().required().description('verification required for using the app'),
+    verificationUrl: joi.string().required().description('the url to PUT a verification with /:id'),
+    logo: joi.string().description('app logo (base64 encoded image)'),
+    url: joi.string().description('the base url for the app'),
+    assignedSponsorships: joi.number().integer().description('number of assigned sponsorships'),
+    unusedSponsorships: joi.number().integer().description('number of unused sponsorships'),
+  }),
+  operation: joi.alternatives().try([
+    joi.object({
+      name: joi.string().valid('Add Connection').required().description('operation name'),
+      _key: joi.string().description('sha256 hash of the operation message used for generating signature'),
```

Comment:
From what I can tell from looking at [index.js line 45](https://github.com/BrightID/BrightID-Node/blob/2e4c8349fdb416c63caf02c50c66d545738371cc/web_services/foxx/brightid5/index.js#L45), `_key` is computed, not supplied. If it were supplied, it would be overwritten.

Reply:
Yes, But this should be optionally here and we can't remove that from joi validation attributes.
This is not only used when users `POST` the operations but in different places in operations life cycle, and this hash is on operations in those places.

Reply:
> Yes, But this should be optionally here and we can't remove that from joi validation attributes.
> This is not only used when users `POST` the operations but in different places in operations life cycle, and this hash is on operations in those places.

Ok, you're talking about operation that is received, not one that is `POST`ed.  You want people to know what is in the object they're receiving and you want to be able to reuse this schema for both.

I think the best approach is two have two schemas: one for `POST`ed operations and one for received operations. This gives the most accurate information to users.

Reply:
Instead of having 2 different schema, I updated things to allow removing `hash` from operations sent to both `brightid` and `apply` services:

https://github.com/BrightID/BrightID-Node/commit/f8df4cc00d0f00fcb5be452f571456517ff6c992
https://github.com/BrightID/BrightID-Node/commit/39051efb59504379879c636f39fffb52e90569fa

Now operations sent to `idchain` do not have `hash` too.
I agree that this help a lot simplifying things for developers want to integrate brightid. :+1: